pkg/storage/stores/shipper/bloomshipper/client.go

```diff
@@ -149,68 +147,72 @@ func (b *BloomClient) PutMeta(ctx context.Context, meta Meta) error {
 	if err != nil {
 		return fmt.Errorf("can not marshal the meta to json: %w", err)
 	}
-	key := externalMetaKey(meta.MetaRef)
+	key := b.Meta(meta.MetaRef).Addr()
 	return b.client.PutObject(ctx, key, bytes.NewReader(data))
 }
 
-func externalMetaKey(ref MetaRef) string {
-	filename := fmt.Sprintf("%s-%d-%d-%x", ref.Bounds.String(), ref.StartTimestamp, ref.EndTimestamp, ref.Checksum)
-	return path.Join(rootFolder, ref.TableName, ref.TenantID, metasFolder, filename)
-}
-
-func findPeriod(configs []config.PeriodConfig, ts model.Time) (config.DayTime, error) {
-	for i := len(configs) - 1; i >= 0; i-- {
-		periodConfig := configs[i]
-		if periodConfig.From.Before(ts) || periodConfig.From.Equal(ts) {
-			return periodConfig.From, nil
-		}
-	}
-	return config.DayTime{}, fmt.Errorf("can not find period for timestamp %d", ts)
-}
+func (b *BloomClient) DeleteMetas(ctx context.Context, refs []MetaRef) error {
+	err := concurrency.ForEachJob(ctx, len(refs), b.concurrency, func(ctx context.Context, idx int) error {
+		key := b.Meta(refs[idx]).Addr()
+		return b.client.DeleteObject(ctx, key)
+	})
 
-func (b *BloomClient) DeleteMeta(ctx context.Context, meta Meta) error {
-	key := externalMetaKey(meta.MetaRef)
-	return b.client.DeleteObject(ctx, key)
+	return err
 }
 
 // GetBlock downloads the blocks from objectStorage and returns the downloaded block
-func (b *BloomClient) GetBlock(ctx context.Context, reference BlockRef) (LazyBlock, error) {
-	readCloser, _, err := b.client.GetObject(ctx, b.Block(reference).Addr())
+func (b *BloomClient) GetBlock(ctx context.Context, ref BlockRef) (BlockDirectory, error) {
+	key := b.Block(ref).Addr()
+	readCloser, _, err := b.client.GetObject(ctx, key)
 	if err != nil {
-		return LazyBlock{}, fmt.Errorf("error while fetching object from storage: %w", err)
+		return BlockDirectory{}, fmt.Errorf("failed to get block from storage: %w", err)
 	}
-	return LazyBlock{
-		BlockRef: reference,
-		Data:     readCloser,
-	}, nil
-}
-
-func (b *BloomClient) PutBlocks(ctx context.Context, blocks []Block) ([]Block, error) {
-	results := make([]Block, len(blocks))
-	err := concurrency.ForEachJob(ctx, len(blocks), b.concurrency, func(ctx context.Context, idx int) error {
-		block := blocks[idx]
-		defer func(Data io.ReadCloser) {
-			_ = Data.Close()
-		}(block.Data)
 
-		var err error
+	path := b.fsResolver.Block(ref).LocalPath()
+	err = extractBlock(readCloser, path, b.logger)
+	if err != nil {
+		return BlockDirectory{}, fmt.Errorf("failed to extract block into directory : %w", err)
+	}
 
-		key := b.Block(block.BlockRef).Addr()
-		_, err = block.Data.Seek(0, 0)
-		if err != nil {
-			return fmt.Errorf("error uploading block file: %w", err)
-		}
+	return NewBlockDirectory(ref, path, b.logger), nil
+}
 
-		err = b.client.PutObject(ctx, key, block.Data)
+func (b *BloomClient) GetBlocks(ctx context.Context, refs []BlockRef) ([]BlockDirectory, error) {
+	// TODO(chaudum): Integrate download queue
```

Comment:
According to the PR desc:
> The blockDownloader is replaced by a simple queue powered by a channel.

Is this TODO outdated? I don't see any queue mechanism at this level, can't we reuse what's implemented in `FetchBlocksWithQueue`?

Reply:
Yes, the TODO is outdated and should be removed.